.github/pull_request_template.md

```diff
@@ -0,0 +1,41 @@
+# Description
+Please include a summary of the changes and the related issue. List any dependencies that are required for this change.
+
+# Linked Items:
+Please include the associated items of this Pull Request using the following syntax to link them for closure on completion of this Pull Request.
+`Closes #<Issue Number>`
+`Fixes #<Issue Number>`
+`Resolves #<Issue Number>`
+
+## Type of change
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+- [ ] Feature improvement
+- [ ] Requires a documentation update
+
+### Feature Additions
+Include pertinent changes as a result of adding the new feature.
+- None
+
+### Bug Fixes
+Include pertinent changes of fixing the bug.
+- None
+
+### Breaking Changes
+Include pertinent changes that qualify as breaking changes. Be sure to include information about the impact to end users and additional information for addressing the break if more than a version transition is necessary for mitigating. 
+- None
+
+### Feature Improvement Changes
+Include pertinent changes that are improvements to existing features.
+- None
+
+# Checklist:
+- [ ] Code follows the style guidelines of this project.
+- [ ] Code is commented (at minimum new code, but if legacy code is missing comments, try to update.)
```

Comment:
I'd favor improving the readability of the code over adding comments as a way to try to describe what's going on so I would suggest we either remove this or update to say that adding comments is needed when it can't be made self-documenting.

Reply:
I'm split on this since just relying on self-documenting code can become very technical and be difficult for newer or less experienced developers whereas a comment can easily provide high-level context to the reader. Since this is an OS project, I'm inclined for us to spend some cycles supporting and maintaining high-level clarifying comments where we can.

Reply:
Middle ground of updating comments as you touch things and adding `TODO` references with issue numbers for code readability refinement and self-documenting improvements? 

Would see that as being two separate check list items.

Reply:
> Can agree to that preference. But that type of change, to me, would likely fall outside of the original scope of the task. Whereas updating comments is easier to cover within scope from what I have usually seen in previous project work.

I don't really see what you mean by "fall outside the original scope of the task". The suggestion is to clarify the comment requirement to say something like "add comments as needed".

> I'm split on this since just relying on self-documenting code can become very technical and be difficult for newer or less experienced developers whereas a comment can easily provide high-level context to the reader. Since this is an OS project, I'm inclined for us to spend some cycles supporting and maintaining high-level clarifying comments where we can.

To clarify, I'm not saying don't add comments at all just opt for code clarity over comments. I've seen it happen where comments are used in place of writing descriptive code and I am strongly opposed to that. That said, could you give an example of what would be more complex when reading self-documenting code (which is known to be up-to-date) over descriptive comments (which can often be out-of-date)? In general, I'd rather spend cycles updating code to be more descriptive (i.e. a descriptive function name with a docstring is 1000x better than a descriptive comment, IMO).

> Middle ground of updating comments as you touch things and adding `TODO` references with issue numbers for flagging areas eligible for code readability refinement and self-documenting improvements?
> 
> Would see that as being two separate check list items.

TODO references will break the linter and hence the pipeline (unless we turn that warning off).

Reply:
> I don't really see what you mean by "fall outside the original scope of the task". The suggestion is to clarify the comment requirement to say something like "add comments as needed" but not require them to be added to all new/updated code which will get superfluous very quickly.

The scope comment was with regards to re-working the code to be more readable and self-documenting, not about modifying the checklist item. My referencing in my response was unclear. 

> TODO references will break the linter and hence the pipeline (unless we turn that warning off).

Good to know, I have not dug through the pylint and flake8 configurations to see what all is currently set. 

Definitely a good discussion piece for me to learn how you all handle referencing issues within the code.

Reply:
Let's remove the code commenting bit for now. I expect reviewers to provide feedback on whether or not the PR content is readable and documented.
```suggestion
```